### PR TITLE
Added Bulk Api functionality to the ToolKit

### PR DESCRIPTION
**Pull request by bradslaterffdc**
Hi, I added Bulk API support to the Toolkit.

The new BulkForceClient supports Insert, Update and Delete operations.
The following features of the Bulk API are not supported yet:
- query jobs
- serial jobs
- jobs that contain attachments
- CSV type jobs

I have written unit tests and functional tests for the `BulkForceClient` in **ForceToolkitForNET** and have written unit tests for the new `BulkServiceHttpClient` in **CommonLibrariesForNET** (But not functional tests as it would require a dependency on the **ForceToolkitForNET** project).

The `BulkServiceHttpClient` and `ServiceHttpClient` could do with a refactor into a `BaseHttpClient`, an `XmlHttpClient` and a `JsonHttpClient` with an extension for the queryMore, etc. 
I didn't do this as I didn't want to refactor the core logic in your project and the pull request is big enough already.

In terms of the project organisation, I did think about putting the bulk stuff in its own package, but that would require a new NuGet package, so I didn't.

I plan to add support for the above areas not currently supported at some point.

I also created two sample projects, one which uses a helper method to simplify things, and one which creates the job/batches step by step and polls for the result.

### CODE REVIEW

**Comment by bradslaterffdc** [comment]
One of the functional tests in ```ForceClientTests``` is failing and i'm not sure why, I haven't even touched that class, and only remove accounts in my functional test that I created, etc.

@dcarroll can you offer any insight?

**Comment by dcarroll** [comment]
I've noticed that sometimes on of the CI builds will fail due to an authentication failure, and if it is rerun will pass.  Not sure why.

File: README.md [hunk 1]
```diff
@@ -164,6 +177,142 @@ foreach (var account in accounts.records)
 }
 ```
 
+### Bulk Sample Code
+
+Below are some simple examples that show how to use the ```BulkForceClient```
+
+**NOTE:** The following features are currently not supported
+
+* CSV data type requests / responses
```
Comment (dcarroll) [inline README.md]: Maybe parenthetically indicate (only json currently support)

**Comment by dcarroll** [comment]
Hey Brad,

Is there support for passing in an xml file?  Also, why did you create a
separate client for bulk?

On Wed, Aug 19, 2015 at 7:40 AM, Brad Slater <notifications@github.com>
wrote:

> One of the functional tests in ForceClientTests is failing and i'm not
> sure why, I haven't even touched that class, and only remove accounts in my
> functional test that I created, etc.
>
> @dcarroll <https://github.com/dcarroll> can you offer any insight?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/developerforce/Force.com-Toolkit-for-NET/pull/136#issuecomment-132623062>
> .
>



-- 
Thanks,
Dave

**Comment by bradslaterffdc** [comment]
Hi @dcarroll ,

I created a separate client in the common libraries to handle XML instead of JSON, and a separate client in the Toolkit to keep things separated and clear. All of the methods on the bulk client basically do one thing, and the helper method CreateJobAndPoll runs through them all in a nice way if you don't want to implement your own polling, etc.

There could be support for passing in an XML file quite easily, you'd just overload CreateJob etc, with methods that took a string and create a method on the HttpClient that didn't try to parse whatever you give it.

I left this out because, as I saw it, the current HttpClient supports JSON, the new BulkHttpClient supports XML, and it didn't feel right to add a method to either that did a straight call via a string. 

In my view, there should be a standard HttpClient that takes and returns strings, and has the default auth settings (overridable if say you need, say ```_httpClient.DefaultRequestHeaders.Add("X-SFDC-Session", accessToken);``` instead of the default bearer auth header.

Then a ```JsonHttpClient``` for the standard stuff with the encoding in overrides and an ```XmlHttpClient``` for the bulk work, that way, things would be nice and clean, and if you wanted a method on the BulkForceClient that just took an XML string, you could just call the standard client.

I didn't do this for now as it would be a major change, and I didn't want to edit your code in my pull request (At least until i'd spoken to you about it first)

Brad

**Comment by dcarroll** [comment]
Thanks for the explanation Brad. I agree with the architecture that you
have described.  It would be nice to have a single point of entry (client)
for interacting with any of the APIs.  As long as there is a single
ForceClient then that would fit the bill.

The problem with going with the Bulk client now, and later weaving bulk
back into the ForceClient is that it is then a breaking change for existing
apps.  I agree it might be a big chunk of work, so I'll take the pull
request, but I don't think I should create a new nuget for it until we
think it is stable.



On Wed, Aug 19, 2015 at 10:33 AM, Brad Slater <notifications@github.com>
wrote:

> Hi @dcarroll <https://github.com/dcarroll> ,
>
> I created a separate client in the common libraries to handle XML instead
> of JSON, and a separate client in the Toolkit to keep things separated and
> clear. All of the methods on the bulk client basically do one thing, and
> the helper method CreateJobAndPoll runs through them all in a nice way if
> you don't want to implement your own polling, etc.
>
> There could be support for passing in an XML file quite easily, you'd just
> overload CreateJob etc, with methods that took a string and create a method
> on the HttpClient that didn't try to parse whatever you give it.
>
> I left this out because, as I saw it, the current HttpClient supports
> JSON, the new BulkHttpClient supports XML, and it didn't feel right to add
> a method to either that did a straight call via a string.
>
> In my view, there should be a standard HttpClient that takes and returns
> strings, and has the default auth settings (overridable if say you need,
> say _httpClient.DefaultRequestHeaders.Add("X-SFDC-Session", accessToken);
> instead of the default bearer auth header.
>
> Then a JsonHttpClient for the standard stuff with the encoding in
> overrides and an XmlHttpClient for the bulk work, that way, things would
> be nice and clean, and if you wanted a method on the BulkForceClient that
> just took an XML string, you could just call the standard client.
>
> I didn't do this for now as it would be a major change, and I didn't want
> to edit your code in my pull request (At least until i'd spoken to you
> about it first)
>
> Brad
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/developerforce/Force.com-Toolkit-for-NET/pull/136#issuecomment-132708562>
> .
>



-- 
Thanks,
Dave

**Comment by wadewegner** [comment]
I would advocate waiting to take the pull request until we can refactor it
appropriately. This can be done in a separate branch.

Great addition, Brad.

**Comment by dcarroll** [comment]
Thanks for weighing in Wade. After sending the comment I was thinking we
create an "unstable" branch and take the PR there.  Thoughts?

On Wed, Aug 19, 2015 at 10:59 AM, Wade Wegner <notifications@github.com>
wrote:

> I would advocate waiting to take the pull request until we can refactor it
> appropriately. This can be done in a separate branch.
>
> Great addition, Brad.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/developerforce/Force.com-Toolkit-for-NET/pull/136#issuecomment-132722234>
> .
>



-- 
Thanks,
Dave

**Comment by bradslaterffdc** [comment]
Hi again @dcarroll @wadewegner 

It sounds like we have the ```HttpClient``` stuff worked out for the common libraries?

My thoughts on a single client are that mainly people will want to just use the 'helper' method I wrote and will rarely implement a step by step approach that uses most methods that the ```BulkForceClient``` exposes. What I would propose is that the ```BulkForceClient``` serves a purpose and keeps all the low level bulk calls from cluttering the main client.

What if we move the 'helper' method, ```RunJobAndPollAsync````, to the standard client? You could also move the constants required out to a root level class too.

That way, most people will just need the standard ```ForceClient``` and only a small subsection of devs will ever need to delve deeper. Kind of like how you can use the ```HttpClient``` in the common libraries if you choose to? If not we could just move it all across, but I think that might be confusing to an end user.

If you want, I can open the pull request to a new unstable branch and fork again afterwards if you'd like me to do the refactor? 

I could also just close the request, do all the refactor, and then open it again to a new branch that you could create for the purpose, all finished?

The advantage of the second way is that you can take it complete in an unstable branch and evaluate the 'finished product' so to speak, and make any changes you feel necessary. 

It would also mean I could continue to work without needing to wait for the merge, etc.

Its getting late over here anyway (UK) let me know what you decide.

Thanks,

Brad

**Comment by dcarroll** [comment]
I'll defer to Wade.

On Wed, Aug 19, 2015 at 12:35 PM, Brad Slater <notifications@github.com>
wrote:

> Hi again @dcarroll <https://github.com/dcarroll> @wadewegner
> <https://github.com/wadewegner>
>
> It sounds like we have the HttpClient stuff worked out for the common
> libraries?
>
> My thoughts on a single client are that mainly people will want to just
> use the 'helper' method I wrote and will rarely implement a step by step
> approach that uses most methods that the BulkForceClient exposes. What I
> would propose is that the BulkForceClient serves a purpose and keeps all
> the low level bulk calls from cluttering the main client.
>
> What if we move the 'helper' method, RunJobAndPollAsync`, to the standard
> client? You could also move the constants required out to a root level
> class too.
>
> That way, most people will just need the standard ForceClient and only a
> small subsection of devs will ever need to delve deeper. Kind of like how
> you can use the HttpClient in the common libraries if you choose to? If
> not we could just move it all across, but I think that might be confusing
> to an end user.
>
> If you want, I can open the pull request to a new unstable branch and fork
> again afterwards if you'd like me to do the refactor?
>
> I could also just close the request, do all the refactor, and then open it
> again to a new branch that you could create for the purpose, all finished?
>
> The advantage of the second way is that you can take it complete in an
> unstable branch and evaluate the 'finished product' so to speak, and make
> any changes you feel necessary.
>
> It would also mean I could continue to work without needing to wait for
> the merge, etc.
>
> Its getting late over here anyway (UK) let me know what you decide.
>
> Thanks,
>
> Brad
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/developerforce/Force.com-Toolkit-for-NET/pull/136#issuecomment-132753400>
> .
>



-- 
Thanks,
Dave

**Comment by bradslaterffdc** [comment]
@wadewegner 

What do you think Wade? We're pretty keen on using this feature in the future, so whatever you decide, I'd like to help as much as possible to get it in a stable state and ready for inclusion in the NuGet package.

**Comment by wadewegner** [comment]
Sorry for the delay. Looking at your code right now.

**Comment by wadewegner** [comment]
After a **short** glance, I think I'd like to see the following:

1. All the Xml-related models/classes moved into the Common library (similar to the Json models today).
2. Refactor ServicesHttpClient into JsonHttpClient and create a comparable XmlHttpClient in Common library? Thinking this one through still.
3. Refactor ForceClient and BulkClient such that the bulk APIs are consolidated into the ForceClient.
4. Functional tests for all the bulk API additions.

Thoughts?

**Comment by bradslaterffdc** [comment]
Hi @wadewegner, 

that all seems sensible to me.

I can get on with this and hopefully be finished in a few days if you want?
There was a bulk functional test, and I removed it as I thought it was making the test build/run fail, although it ran fine locally.
On closer inspection, this seems to be a bit of a lottery anyway.

Brad

**Comment by bradslaterffdc** [comment]
@wadewegner @dcarroll 

Hi,

I completed the refactor (sorry about the delay, I've been really busy these past few weeks) and merged upstream. Everything is working / all tests run locally.

The only issue is my functional test for bulk fails when running on travis. 
I'm going to debug this now / tomorrow.

Just wanted to let you know I was done (ish)

**Comment by bradslaterffdc** [comment]
@wadewegner @dcarroll 

All bugs fixed in the project after much messing about.
As long as your happy with the refactor you can merge now and I won't complain :)

* The refactor of the HttpServices was really worthwhile and the JSON service actually looks to have benefited the most from the changes.

* I'm still not sure about merging all the new methods into the force client, but it was your choice so I did it.

* All the models are now in the common libraries too.

* There is a functional test that runs through adding and deleting batches, etc. this test uses the main 'helper' method, and as such runs through all the other low level batch methods.

When this is merged, what are your thoughts on a new NuGet package?

Brad

PS. I had to remove a load of test accounts from the org in the end to get the build to run. Kept getting Storage Limit Exceeded. Maybe we should change the tests in the future to tidy up after themselves?

**Comment by wadewegner** [comment]
Thanks for the contribution. I'll set aside some time to review, but note that it may take a wee bit. =)

**Comment by bradslaterffdc** [comment]
Ok, thanks very much for taking the time. I understand it might take a while, its quite a big change.

Brad

**Comment by bradslaterffdc** [comment]
Hi @wadewegner, @dcarroll

Have you made any progress with this?
We have a dev team interested in evaluating the toolkit for a live project, and one of their requirements was bulk support.
They can use my fork of the repo for now, but have indicated they would like a NuGet package including the features sooner or later.

I know its a busy time right now anyway, so no pressure :)

**Comment by twhitbeck** [comment]
Would love to see this feature merged.

**Comment by dcarroll** [comment]
I should have some time to look at this in the afternoon here in Cali.

**Comment by haicamha** [comment]
Hi Bradslaterffdc,
   I used your library to do the bulk api, but noticed that it doesn't support the Upsert. If possible, could you please add this feature into the library. 

   I have implemented this feature (Upsert) by adding a new field to the class JobInfo as shown below. And when doing the Upsert, I have to specify the value of this new field. Usually it is "Id", but sometimes the object's primary key is not Id. It is a different field.

   In the class JobInfo, I add a field called ExternalId right before the ContentType field. Please note that the order of the field is very important, otherwise it won't work.

        [XmlElement(ElementName = "externalIdFieldName")]
        public string ExternalId { get; set; }

        [XmlElement(ElementName = "contentType")]
        public string ContentType { get; set; }

Thanks & Regards,
Hai

**Comment by mp3tobi** [comment]
Will this pull request ever merged?
I think the bulk API could be very helpful for customers who have over 100.000 products and want it syncronize via this API. :)

**Comment by amhed** [comment]
+1 on Merging this. I need bulk functionality for a project and would make life much easier

**Comment by wadewegner** [comment]
I've started the merge. Holy cow. Lots here. :)

**Comment by wadewegner** [comment]
Merging this is going to be extremely difficult. There's a real risk that updates will be lost. Additionally, I'm finding that the new internals (specifically `BaseHttpClient`) have lost the ability to pass in the type. Everything's changed to string, which breaks most everything. Not sure how big a fix this will require.

**Comment by wadewegner** [comment]
Okay, update. I've gotten this entirely merged and tests are passing. However, I know that we're missing updates (GZIP, for example).

So, I'm going to push the updates and but yet merge. We can work through the updates.

Thanks for your patience everyone.

**Comment by wadewegner** [comment]
Note: looks like my push went into this branch rather than updating this pull request: https://github.com/developerforce/Force.com-Toolkit-for-NET/tree/bradslaterffdc-master

**Comment by wadewegner** [comment]
Incidentally, I was able to update this branch to work with Travis CI finally.

**Comment by jon-freed** [comment]
@wadewegner , so does your July 29 post mean this pull request should be closed because, in essence, it was done?  Or what is the plan for that [bradslaterffdc-master branch](https://github.com/developerforce/Force.com-Toolkit-for-NET/tree/bradslaterffdc-master)?

I took a look at the [branches](https://github.com/developerforce/Force.com-Toolkit-for-NET/branches) and [compared](https://github.com/developerforce/Force.com-Toolkit-for-NET/compare/bradslaterffdc-master) that branch to master, and GitHub seems optimistic that the bradslaterffdc-master branch could be automatically merged with the master branch.  Is that the plan?  Thanks!

**Comment by juan-cs** [comment]
I would also love this PR to be merged! Thanks everyone for the great job!

**Comment by naicigam** [comment]
I would also love this PR to be merged! Thanks everyone for the great job!

**Comment by wadewegner** [comment]
Well, look at this! It's a 2017 miracle. Merged this PR ***and*** fixed the travis-ci build. Someone pinch me!

I'm going to look at accepting a few additional pull requests. Then I'll look into updating NuGet.

Note: I'm going to need some help updating all the documentation and sample apps.

**Comment by naicigam** [comment]
Awesome! Congratulations!

**Comment by yehia2amer** [comment]
Hello @wadewegner is there any plans to support CSVs in the Bulk Api ?

**Comment by alcidesfmelof** [comment]
BulkForceClient still in beta version? Any update on that? 

Keep the good work guys!!

**Pull request merged**